examples/moving_least_squares/moving_least_squares.cpp

```diff
@@ -0,0 +1,343 @@
+// (http://dx.doi.org/10.1016/j.jcp.2015.11.055)
+
+#include <ArborX.hpp>
+
+#include <Kokkos_Core.hpp>
+
+#include <limits>
+
+using ExecutionSpace = Kokkos::DefaultExecutionSpace;
+using MemorySpace = ExecutionSpace::memory_space;
+
+struct RBFWendland_0
+{
+  KOKKOS_INLINE_FUNCTION float operator()(float x)
+  {
+    x /= _radius;
+    return (1. - x) * (1. - x);
+  }
+
+  float _radius;
+};
+
+struct MVPolynomialBasis_3D
+{
+  static constexpr std::size_t size = 4;
+
+  KOKKOS_INLINE_FUNCTION Kokkos::Array<float, size>
+  operator()(ArborX::Point const &p) const
+  {
+    return {{1., p[0], p[1], p[2]}};
+  }
+};
+
+// Function to approximate
+KOKKOS_INLINE_FUNCTION float manufactured_solution(ArborX::Point const &p)
+{
+  return p[2] + p[0];
+}
+
+int main(int argc, char *argv[])
+{
+  Kokkos::ScopeGuard guard(argc, argv);
+  constexpr std::size_t num_neighbors = 5;
+  constexpr std::size_t cube_side = 4;
+  constexpr std::size_t source_points_num = cube_side * cube_side * cube_side;
+  constexpr std::size_t target_points_num = 4;
+
+  auto source_points = Kokkos::View<ArborX::Point *, MemorySpace>(
+      Kokkos::view_alloc(Kokkos::WithoutInitializing, "MLS_EX::source_points"),
+      source_points_num);
+  auto target_points = Kokkos::View<ArborX::Point *, MemorySpace>(
+      Kokkos::view_alloc(Kokkos::WithoutInitializing, "MLS_EX::target_points"),
+      target_points_num);
+  auto target_points_host = Kokkos::create_mirror_view(target_points);
+
+  // Generate source points (Organized within a [-10, 10]^3 cube)
+  Kokkos::parallel_for(
+      "MLS_EX::source_points_init",
+      Kokkos::MDRangePolicy<ExecutionSpace, Kokkos::Rank<3>>(
+          {0, 0, 0}, {cube_side, cube_side, cube_side}),
+      KOKKOS_LAMBDA(int const i, int const j, int const k) {
+        source_points(i * cube_side * cube_side + j * cube_side + k) =
+            ArborX::Point{20.f * (float(i) / (cube_side - 1) - .5f),
+                          20.f * (float(j) / (cube_side - 1) - .5f),
+                          20.f * (float(k) / (cube_side - 1) - .5f)};
+      });
+
+  // Generate target points
+  target_points_host(0) = ArborX::Point{0.f, 0.f, 0.f};
+  target_points_host(1) = ArborX::Point{5.f, 5.f, 5.f};
+  target_points_host(2) = ArborX::Point{-5.f, 5.f, 3.f};
+  target_points_host(3) = ArborX::Point{1.f, -3.3f, 7.f};
+  Kokkos::deep_copy(target_points, target_points_host);
+
+  // Organize source points as tree
+  ArborX::BVH<MemorySpace> source_tree(ExecutionSpace{}, source_points);
+
+  // Create the queries
+  // For each target point we query the closest source points
+  auto queries = Kokkos::View<ArborX::Nearest<ArborX::Point> *, MemorySpace>(
+      "MLS_EX::queries", target_points_num);
+  Kokkos::parallel_for(
+      "MLS_EX::make_queries",
+      Kokkos::RangePolicy<ExecutionSpace>(0, target_points_num),
+      KOKKOS_LAMBDA(int const i) {
+        queries(i) = ArborX::nearest(target_points(i), num_neighbors);
+      });
```

Comment:
That's how it was done in DTK, so I copied. But I agree that using `AccessTraits` would be a more "correct" way to do it.

Reply:
> That's how it was done in DTK

We've come a long way since DTK. DTK is pretty outdated in the way it uses ArborX. It is a good starting point, but we want and should move more towards a better and more efficient use of ArborX.